File: mock-facility-management/src/main/java/de/vkb/dojo/es/facilityManagement/services/events/EventAggregatorResult.java
package de.vkb.dojo.es.facilityManagement.services.events;


import de.vkb.dojo.es.facilityManagement.model.event.Event;
import de.vkb.dojo.es.facilityManagement.model.feedback.Feedback;
import de.vkb.dojo.es.facilityManagement.model.feedback.SuccessFeedback;

public class EventAggregatorResult<E extends Event, A> {
    private final E event;
    private final A aggregate;
    private final Feedback feedback;

    public EventAggregatorResult(E event, A aggregate) {
        this.event = event;
        this.aggregate = aggregate;
        this.feedback = new SuccessFeedback(event.getReference());
    }

    public EventAggregatorResult(E event, Feedback feedback) {
        this.event = event;
        this.aggregate = null;
        this.feedback = feedback;
    }

    public EventAggregatorResult(E event) {
        this.event = event;
        this.aggregate = null;
        this.feedback = new SuccessFeedback(event.getReference());
    }

    public E getEvent() {
        return event;
    }

    public A getAggregate() {
        return aggregate;
    }

    public Feedback getFeedback() {
        return feedback;
    }
}

File: mock-human-resources/src/main/kotlin/de/vkb/dojo/es/humanResources/services/events/EventAggregatorResult.kt
package de.vkb.dojo.es.humanResources.services.events

import de.vkb.dojo.es.humanResources.model.event.Event
import de.vkb.dojo.es.humanResources.model.feedback.Feedback
import de.vkb.dojo.es.humanResources.model.feedback.SuccessFeedback

data class EventAggregatorResult<E: Event, A>(
    val event: E,
    val aggregate: A?,
    val feedback: Feedback
)

fun <E: Event, A> EventAggregatorResult(event: E, feedback: Feedback): EventAggregatorResult<E, A> = EventAggregatorResult(event, null, feedback)
fun <E: Event, A> EventAggregatorResult(event: E, aggregate: A): EventAggregatorResult<E, A> = EventAggregatorResult(event, aggregate, SuccessFeedback(event.reference))
